src/logic/function/basic.lean

```diff
@@ -540,14 +540,23 @@ along a function `f : α → β` to a function `β → γ`,
 by using the values of `g` on the range of `f`
 and the values of an auxiliary function `e' : β → γ` elsewhere.
 
-Mostly useful when `f` is injective. -/
+Mostly useful when `f` is injective,
+more generally when `f a  f b → g a = g b`. -/
 def extend (f : α → β) (g : α → γ) (e' : β → γ) : β → γ :=
 λ b, if h : ∃ a, f a = b then g (classical.some h) else e' b
 
 lemma extend_def (f : α → β) (g : α → γ) (e' : β → γ) (b : β) [decidable (∃ a, f a = b)] :
   extend f g e' b = if h : ∃ a, f a = b then g (classical.some h) else e' b :=
 by { unfold extend, congr }
 
+lemma extend_apply_of_unique (g : α → γ) (e' : β → γ)
+  (hf : ∀ ⦃a b : α⦄, f a = f b → g a = g b) (a : α) :
+  extend f g e' (f a) = g a :=
+begin
+  simp only [extend_def, dif_pos, exists_apply_eq_apply],
+  exact hf (classical.some_spec (exists_apply_eq_apply f a)),
+end
+
 @[simp] lemma extend_apply (hf : injective f) (g : α → γ) (e' : β → γ) (a : α) :
   extend f g e' (f a) = g a :=
 begin
```

Comment:
I think it makes sense to golf this using the new result. It makes it clear that the new one is more general.

Reply:
I have golfed two things, one this morning, one this afternoon, 
and also added : 
`function.factors_through`
and 
`function.factors_through_of_is_injective`
but I don't see it appear here…

Reply:
OK, now it does appear…